packages/headless/src/features/insight-user-actions/insight-user-actions-preprocessing.test.ts

```diff
@@ -0,0 +1,645 @@
+        searchHub: 'in-product-help',
+        document: {title: 'title', uriHash: 'TtnKwc0Lo2GY9WAi'},
+      };
+
+      // Added 1 min to action timestamp, so it is within 30 mins of the previous action
+      const previousEndDateTime = JSON.stringify(
+        createRelativeDate(firstSessionDate, 1, 0).getTime()
+      );
+
+      const isSameSession = isActionWithinSessionThreshold(
+        action,
+        previousEndDateTime
+      );
+      expect(isSameSession).toEqual(true);
+    });
+
+    it('should return false if the action is not within 30mins of the previous', async () => {
+      const action = {
+        actionType: 'CLICK' as UserActionType,
+        timestamp: firstSessionDate.getTime().toString(),
+        eventData: {},
+        searchHub: 'in-product-help',
+        document: {title: 'title', uriHash: 'TtnKwc0Lo2GY9WAi'},
+      };
+
+      // Added 1 hour to action timestamp, so it is not within 30 mins of the previous action
+      const previousEndDateTime = JSON.stringify(
+        createRelativeDate(firstSessionDate, 60, 0).getTime()
+      );
+
+      const isSameSession = isActionWithinSessionThreshold(
+        action,
+        previousEndDateTime
+      );
+      expect(isSameSession).toEqual(false);
+    });
+  });
+
+  describe('#filterTimelineActions', () => {
+    it('should properly filter out the timeline of the actions that are to be excluded and of empty searches', async () => {
+      const actionsToExclude = ['useless_event'];
+      const mockRawActions = [...fakeActions].slice(0, 8);
+      const mappedAndSortedActions =
+        mapAndSortActionsByMostRecent(mockRawActions);
+      const ticketCreationDate = JSON.stringify(caseCreationDate.getTime());
+      const sessionsTimeline = splitActionsIntoTimelineSessions(
+        mappedAndSortedActions,
+        ticketCreationDate
+      );
+
+      const filteredTimeline = filterTimelineActions(
+        sessionsTimeline,
+        actionsToExclude
+      );
+
+      // Checks that no sessions in each section of the timeline contains any of the actions to exclude
+      actionsToExclude.forEach((actionType) => {
+        filteredTimeline.precedingSessions.forEach((session) => {
+          expect(
+            session.actions.every((action) => action.actionType !== actionType)
+          ).toEqual(true);
+          expect(
+            session.actions.every((action) => action.query !== '')
+          ).toEqual(true);
+        });
+
+        filteredTimeline.followingSessions.forEach((session) => {
+          expect(
+            session.actions.every((action) => action.actionType !== actionType)
+          ).toEqual(true);
+          expect(
+            session.actions.every((action) => action.query !== '')
+          ).toEqual(true);
+        });
+
+        filteredTimeline.session?.actions.forEach((action) => {
+          expect(action.actionType !== actionType).toEqual(true);
+          expect(action.query !== '').toEqual(true);
```

Comment:
for me that's not the best way to perform this test, cause there is a lot of logic to perform the comparison which is not ideal for a test, see comment above for suggestion.

Reply:
I can do a deep equal here no worries